File: September-LeetCoding-Challenge/06-Image-Overlap/Image-Overlap.cs
public class Solution {
    public int LargestOverlap(int[][] A, int[][] B) {
        if(A == null || A.Length ==0 || B==null || B.Length==0) return 0;
        
        int rows = A.Length, cols = A[0].Length;
        List<int[]> al = new List<int[]>(), bl = new List<int[]>();

        for (int i = 0; i < rows; i++)
        {
            for (int j = 0; j < cols; j++)
            {
                if(A[i][j] == 1) al.Add(new int[]{i, j});
                if(B[i][j] == 1) bl.Add(new int[]{i, j });
            }
        }

        var dict = new Dictionary<string, int>();
        foreach (int[] a in al)
        {
            foreach (int[] b in bl)
            {
                var s = (a[0] - b[0]) + " " + (a[1] - b[1]);
                if (dict.ContainsKey(s))
                {
                    dict[s]++;
                }
                else
                {
                    dict.Add(s, 1);
                }
            }
        }
        
        int c = 0;
        foreach(var v in dict.Values){
            c = Math.Max(c, v);
        }
        return c;
    }
}

File: September-LeetCoding-Challenge/06-Image-Overlap/Image-Overlap.rb
def largest_overlap(a, b)

    # c'est une solution que je n'ai pas tout comprise
    # pk [max_overlaps, shift_count(x, y, a, b)].max  et [max_overlaps, shift_count(x, y, b, a)].max suffisent ??
    max_overlaps = 0
    
    len = a.length
    
    0.upto(len-1) do |y|
        0.upto(len-1) do |x|
            max_overlaps = [max_overlaps, shift_count(x, y, a, b)].max
            max_overlaps = [max_overlaps, shift_count(x, y, b, a)].max
        end
    end
    max_overlaps
    
end
        
        
        
def shift_count(x, y, m, r)
   len = m.length
   count = 0
   r_row = 0
    
    y.upto(len - 1) do |m_row|
       r_col = 0
        
            x.upto(len - 1) do |m_col|
                count += 1 if m[m_row][m_col] == 1 &&  m[m_row][m_col] == r[r_row][r_col]
                r_col += 1
            end
        
        r_row +=1
    end
    count
end